data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/ScanObjectWorker.java

```diff
@@ -145,8 +153,11 @@ private void startProcessingObject(final int waitTimeMillis) {
                     if (result == true) {
                         sourceCoordinator.completePartition(objectToProcess.get().getPartitionKey(), true);
                         waitingForAcknowledgements.forEach(s3ObjectDeleteWorker::deleteS3Object);
+                    } else {
+                        sourceCoordinator.giveUpPartition(objectToProcess.get());
                     }
                 }, ACKNOWLEDGEMENT_SET_TIMEOUT);
+                partitions.remove(objectToProcess.get());
```

Comment:
I think this is meant to be in the callback?

Reply:
Yes. Thanks for catching it.